src/tpm/tpm2_interface.c

```diff
@@ -59,8 +59,9 @@ int init_tpm2_connection(TSS2_SYS_CONTEXT ** sapi_ctx)
     // If SAPI initialization fails: 
     //   - sapi_ctx is freed by init_sapi()
     //   - tcti_ctx must still be cleaned up
-    Tss2_Tcti_Finalize(tcti_ctx);
-    free(tcti_ctx);
+    Tss2_TctiLdr_Finalize(&tcti_ctx);
+    if (tcti_ctx != NULL) free(tcti_ctx);
```

Comment:
I'm not quite sure what's supposed to be happening here. Are we only ```free```ing ```tcti_ctx``` if it's not ```NULL```, but we always want it to be ```NULL``` when the function returns? I'd prefer not to use single line ```if``` statements (or ```if``` statements without braces) just so it's always clear the line after the conditional is intentionally not included.

Reply:
I think I was, as you say, checking to see if it is non-NULL (allocated) and only free it in that case. Then I am re-initializing the pointer to free, unconditionally. In this case, the free() and NULL assignment statements should have probably have both been only conditionally executed under the if. I agree that if statement without braces are not the best idea. I also think that, because calling free() on a potentially NULL pointer, though, should not be a problem, so we probably shouldn't be complicating things by making free() calls conditional.

I also re-investigated the behavior of  the Tss2_TctiLdr_Finalize() and Tss2_Sys_Finalize() functions.

- Tss2_TctiLdr_Finalize() destroys an instance of a TCTI context instantiated by the Tss2_TctiLdr_Initialize() function. It also frees any resources associated with loading the required TCTI library. The input parameter is a double pointer to a TCTI context. When successfully finalized, the provided  reference will be set to NULL by the function. Passing a potentially NULL pointer parameter as input seems safe (param is checked). Therefore, subsequent free() and/or NULL assignments should not be required.

- Tss2_Sys_Finalize() will not free caller's pre-allocated memory, so the subsequent free() call for *sapi_ctx is indeed required.

Commit [6024229](https://github.com/NationalSecurityAgency/kmyth/pull/195/commits/60242295c13090b588713eb68e7218d0ba22d43d) contains the changes to the SAPI/TCTI context clean-up code instances.

The conditional free() as a one-line if statement is pretty prevalent in the kmyth code. Commit [2f41e5f](https://github.com/NationalSecurityAgency/kmyth/pull/195/commits/2f41e5f97cde66e23b7ca1a6db5265aa106df720) attempts to address these in a consistent manner.